modules/openapi-generator/src/main/resources/ruby-client/faraday_api_client.mustache

```diff
@@ -0,0 +1,392 @@
+=begin
+{{> api_info}}
+=end
+
+require 'date'
+require 'faraday'
+require 'json'
+require 'logger'
+require 'tempfile'
+
+module {{moduleName}}
+  class ApiClient
+    # The Configuration object holding settings to be used in the API client.
+    attr_accessor :config
+
+    # Defines the headers to be used in HTTP requests of all API calls by default.
+    #
+    # @return [Hash]
+    attr_accessor :default_headers
+
+    # Initializes the ApiClient
+    # @option config [Configuration] Configuration for initializing the object, default to Configuration.default
+    def initialize(config = Configuration.default)
+      @config = config
+      @user_agent = "{{#httpUserAgent}}{{{.}}}{{/httpUserAgent}}{{^httpUserAgent}}OpenAPI-Generator/#{VERSION}/ruby{{/httpUserAgent}}"
+      @default_headers = {
+        'Content-Type' => 'application/json',
+        'User-Agent' => @user_agent
+      }
+    end
+
+    def self.default
+      @@default ||= ApiClient.new
+    end
+
+    # Call an API with given options.
+    #
+    # @return [Array<(Object, Integer, Hash)>] an array of 3 elements:
+    #   the data deserialized from response body (could be nil), response status code and response headers.
+    def call_api(http_method, path, opts = {})
+      connection = Faraday.new(:url => config.base_url) do |conn|
+        conn.basic_auth(config.username, config.password)
+        if opts[:header_params]["Content-Type"] == "multipart/form-data"
+            conn.request :multipart
+            conn.request :url_encoded
+        end
+        conn.adapter(Faraday.default_adapter)
+      end
+      begin
+          response = connection.public_send(http_method.to_sym.downcase) do |req|
+              build_request(http_method, path, req, opts)
+          end
+
+          if @config.debugging
+            @config.logger.debug "HTTP response body ~BEGIN~\n#{response.body}\n~END~\n"
+          end
+
+          unless response.success?
+            if response.status == 0
+              # Errors from libcurl will be made visible here
+              fail ApiError.new(:code => 0,
+                                :message => response.return_message)
+            else
+              fail ApiError.new(:code => response.status,
+                                :response_headers => response.headers,
+                                :response_body => response.body),
+                   response.reason_phrase
+            end
+          end
+      rescue Faraday::TimeoutError
```

Comment:
I'm torn about making all failign cases the same Exception class, it may be difficult to distinguish among errors if that's what the caller wants to do. Also here you probably want to deal with Net::ReadTimeout and Faraday::ConnectionFailed